In [2]:
import json
import plotly.express as px
import pandas as pd
from datetime import datetime
import numpy as np
import json
import warnings
import geojson
import plotly.graph_objs as go
from prophet import Prophet
warnings.filterwarnings("ignore")

## Data Preparation

### Our World in Data

In [78]:
# Sort dataframe by date
df = df.sort_values(by="date")

# Convert date to pd datetime object
df['date'] = pd.to_datetime(df.date,format='%Y-%m-%d')

# Extracting day month year from date
df['month'] = df['date'].dt.month
df['year'] = df['date'].dt.year
df['month_year'] = df['date'].dt.to_period('M')

In [79]:
# Since there is no data available before, we will shorten the dataframe 
df = df[df.date > "2020-01-22"]

# Aggregate the data on a monthly basis for each country 
df = df.groupby(['iso_code','month_year','month','year','location','continent']).sum().reset_index()

# Sort dataframe by month
df = df.sort_values(by="month_year")

# To extract only data regarding Germany
germany = df[df.location == 'Germany']

### DIVI

Anreicherung der DIVI Daten um die Infos der Landkreise sowie hinzufügen des Bundeslandes als Name.
Die Daten werden aus einer Geojson Datei gewonnen.

In [80]:
path_to_file = 'landkreise.geojson'
with open(path_to_file) as f:
    land_geojson= geojson.load(f)

In [81]:
land_geojson['features'][0]['properties']

{'ADE': 4,
 'GF': 4,
 'BSG': 1,
 'RS': '01001',
 'AGS': '01001',
 'SDV_RS': '010010000000',
 'GEN': 'Flensburg',
 'BEZ': 'Kreisfreie Stadt',
 'IBZ': 40,
 'BEM': '--',
 'NBD': 'ja',
 'SN_L': '01',
 'SN_R': '0',
 'SN_K': '01',
 'SN_V1': '00',
 'SN_V2': '00',
 'SN_G': '000',
 'FK_S3': 'R',
 'NUTS': 'DEF01',
 'RS_0': '010010000000',
 'AGS_0': '01001000',
 'WSK': '2008/01/01',
 'DEBKG_ID': 'DEBKGDL20000002R'}

AGS steht für den Gemeindeschlüssel im DIVI Datensatz. Desweiteren enthält die Geojson noch den Name des Landkreises.

In [83]:
# Alle Landkreise werden in einer Liste abgespeichert
codes = divi['gemeindeschluessel'].astype('int').drop_duplicates()
codes = codes.tolist()

dict_codes = {}

# Da die Daten in der DIVI Datei teilweise 4- bzw. 5 stellig vorliegen, sollen diese 5stellig standardisiert werden. Bsp. 1001 wird zu 01001
# Die alten Codes aus der DIVI Datei werden mit den standardisierten in einem Dictionary abgespeichert 

for x in range(0,len(codes)):
   
    code = codes[x]

    if code < 10000:
        dict_codes[code] = "0"+str(code)
    else:
        dict_codes[code] = str(code)
dict_codes

{1001: '01001',
 1002: '01002',
 1003: '01003',
 1004: '01004',
 1051: '01051',
 1053: '01053',
 1054: '01054',
 1055: '01055',
 1056: '01056',
 1057: '01057',
 1058: '01058',
 1059: '01059',
 1060: '01060',
 1061: '01061',
 1062: '01062',
 2000: '02000',
 3101: '03101',
 3102: '03102',
 3103: '03103',
 3151: '03151',
 3153: '03153',
 3154: '03154',
 3155: '03155',
 3157: '03157',
 3158: '03158',
 3159: '03159',
 3241: '03241',
 3251: '03251',
 3252: '03252',
 3254: '03254',
 3255: '03255',
 3256: '03256',
 3257: '03257',
 3351: '03351',
 3352: '03352',
 3353: '03353',
 3354: '03354',
 3355: '03355',
 3356: '03356',
 3357: '03357',
 3358: '03358',
 3359: '03359',
 3360: '03360',
 3361: '03361',
 3401: '03401',
 3402: '03402',
 3403: '03403',
 3404: '03404',
 3405: '03405',
 3451: '03451',
 3452: '03452',
 3453: '03453',
 3454: '03454',
 3455: '03455',
 3456: '03456',
 3457: '03457',
 3458: '03458',
 3459: '03459',
 3460: '03460',
 3461: '03461',
 3462: '03462',
 4011: '04011',
 4012: '

In [84]:
# Der neue Standardisierte Gemeindeschlüssel wird aus dem Dictionary über den bisherigen identifiziert und als neue Spalte dem Datensatz angehängt. 
divi["id"] = divi["gemeindeschluessel"].apply(lambda x: dict_codes[x])

# Die Bundesländer werden ebenfalls in einem Dictionary mit dem jeweiligen Bundesländercode versehen
bundesland = ["Schleswig-Holstein","Hamburg","Niedersachsen","Bremen","Nordrhein-Westfalen","Hessen","Rheinland-Pfalz","Baden-Württemberg","Bayern","Saarland","Berlin","Brandenburg","Mecklenburg-Vorpommern","Sachsen","Sachsen-Anhalt","Thüringen"]
code = range(1,17)
bund_code = dict(zip(code, bundesland))

# Der Name des Bundeslandes wird über den Code im Dictionary identifiziert
divi["bundesland_name"] = divi["bundesland"].apply(lambda x: bund_code[x])

# Die Landkreise werden mit dem Gemeindeschlüssel in einem Dictionary gespeichert
dict_landkreise = {}
for feature in land_geojson['features']:
    dict_landkreise[feature["properties"]["AGS"]] = feature["properties"]["GEN"]

# Der Landkreisname wird über die jeweilige standardisierte ID identifiziert und dem Datensatz hinzugefügt
divi["name"] = divi["id"].apply(lambda x: dict_landkreise[x])

In [85]:
# Datensatz wird auf die benötigten Spalten reduziert
divi = divi[['date','id', 'bundesland_name', 'name','faelle_covid_aktuell',
       'faelle_covid_aktuell_invasiv_beatmet', 'betten_frei', 'betten_belegt']]
divi.head()

date     id     bundesland_name          name  faelle_covid_aktuell  \
0  2020-04-24  01001  Schleswig-Holstein     Flensburg                     1   
1  2020-04-24  01002  Schleswig-Holstein          Kiel                     4   
2  2020-04-24  01003  Schleswig-Holstein        Lübeck                     4   
3  2020-04-24  01004  Schleswig-Holstein    Neumünster                     1   
4  2020-04-24  01051  Schleswig-Holstein  Dithmarschen                     3   

   faelle_covid_aktuell_invasiv_beatmet  betten_frei  betten_belegt  
0                                     1           40             38  
1                                     3           93            108  
2                                     4          109            108  
3                                     1           17             21  
4                                     1           47             25

### RKI

In [86]:
# Dataset wird gefiltert
rki = rki[rki.LandkreisId_Impfort != "u"]

# Landkreis Ids werden als integer markiert
rki.LandkreisId_Impfort = rki.LandkreisId_Impfort.astype('int')

#### Anreicherung der Daten mit 5-stelligen Landkreis Ids

Die Landkreis Ids ohne Duplikate in einer List gespeichert.

In [91]:
lk = rki.LandkreisId_Impfort.drop_duplicates().tolist()
lk[:5]

[1003, 1053, 1055, 1056, 1057]

Da die Landkreise nicht alle 5-stellig vorliegen, wird ein dictionary mit dem 4-stelligen und dem 5-stelligen Wert gebildet. Den 4-stelligen Zahlen wird an der ersten Stelle eine 0 hinzugefügt.

So entsteht bspw. das Key-Value Paar:
1003:01003

Ist die Zahl bereits fünfstellig, so wird der Wert der Liste gespiegelt.
12002:12002

In [92]:
dict_ids = {}
for x in lk:
    if x < 10000:
        dict_ids[x] = "0"+str(x)
    else:
        dict_ids[x] = str(x)
dict_ids[1003], dict_ids[10042]

('01003', '10042')

Die neue Id wird über die Column LandkreisId_Impfort, deren Wert als Key im Dictionary vorliegt, als neue Spalte im Datensatz angelegt.

In [93]:
rki['Id'] = rki["LandkreisId_Impfort"].apply(lambda x: dict_ids[x])
# Die alte Id wird aus dem Datensatz gelöscht
rki = rki.drop('LandkreisId_Impfort',axis=1)
# Da keine Id mit 17000 in der Realität existiert, wird diese aus dem Datensatz entfernt
rki = rki[rki.Id != '17000']

Da nun die vollständigen Ids der Landkreise vorliegen, können die Daten nun mit Informationen aus der geojson angereichert werden.

In [94]:
rki["Name"] = rki["Id"].apply(lambda x: dict_landkreise[x])
rki.head()

In [95]:
rki.head()

Impfdatum Altersgruppe  Impfschutz  Anzahl     Id                 Name
0  2020-12-27        18-59           1       9  01003               Lübeck
1  2020-12-27          60+           1      19  01003               Lübeck
2  2020-12-27        18-59           1      13  01053  Herzogtum Lauenburg
3  2020-12-27          60+           1     200  01053  Herzogtum Lauenburg
4  2020-12-27        18-59           1      20  01055          Ostholstein

In [96]:
divi

date     id     bundesland_name                  name  \
0       2020-04-24  01001  Schleswig-Holstein             Flensburg   
1       2020-04-24  01002  Schleswig-Holstein                  Kiel   
2       2020-04-24  01003  Schleswig-Holstein                Lübeck   
3       2020-04-24  01004  Schleswig-Holstein            Neumünster   
4       2020-04-24  01051  Schleswig-Holstein          Dithmarschen   
...            ...    ...                 ...                   ...   
249533  2022-01-12  16073           Thüringen   Saalfeld-Rudolstadt   
249534  2022-01-12  16074           Thüringen  Saale-Holzland-Kreis   
249535  2022-01-12  16075           Thüringen      Saale-Orla-Kreis   
249536  2022-01-12  16076           Thüringen                 Greiz   
249537  2022-01-12  16077           Thüringen      Altenburger Land   

        faelle_covid_aktuell  faelle_covid_aktuell_invasiv_beatmet  \
0                          1                                     1   
1                          4                                     3   
2                          4                                     4   
3                          1                                     1   
4                          3                                     1   
...                      ...                                   ...   
249533                     6                                     3   
249534                     1                                     0   
249535                     2                                     1   
249536                     2                                     1   
249537                     4                                     1   

        betten_frei  betten_belegt  
0                40             38  
1                93            108  
2               109            108  
3                17             21  
4                47             25  
...             ...            ...  
249533            6             16  
249534            4              4  
249535            0              4  
249536            2              8  
249537            8             24  

[249538 rows x 8 columns]

Die RKI Daten werden mit den DIVI Daten gemerged und darum ergänzt.

In [97]:
rki = pd.merge(rki,divi[['id','bundesland_name']].drop_duplicates(), left_on=["Id"], right_on=["id"], how="left")
rki = rki.drop('id',axis=1)
rki.head()

Impfdatum Altersgruppe  Impfschutz  Anzahl     Id                 Name  \
0  2020-12-27        18-59           1       9  01003               Lübeck   
1  2020-12-27          60+           1      19  01003               Lübeck   
2  2020-12-27        18-59           1      13  01053  Herzogtum Lauenburg   
3  2020-12-27          60+           1     200  01053  Herzogtum Lauenburg   
4  2020-12-27        18-59           1      20  01055          Ostholstein   

      bundesland_name  
0  Schleswig-Holstein  
1  Schleswig-Holstein  
2  Schleswig-Holstein  
3  Schleswig-Holstein  
4  Schleswig-Holstein

In [98]:
# Der Impfschutz ist als Integer hinterlegt und wird nun mit einem Wert versehen, der selbstsprechend ist
rki.loc[rki.Impfschutz == 1, "Impfschutz"] = "Erstimpfung"
rki.loc[rki.Impfschutz == 2, "Impfschutz"] = "Zweitimpfung"
rki.loc[rki.Impfschutz == 3, "Impfschutz"] = "Boosterimpfung"

#Die Spalten werden umbenannt
rki.columns = ['Datum','Altersgruppe','Impfschutz','Anzahl','Id','Landkreis','Bundesland']
rki.head()

Datum Altergruppe   Impfschutz  Anzahl     Id            Landkreis  \
0  2020-12-27       18-59  Erstimpfung       9  01003               Lübeck   
1  2020-12-27         60+  Erstimpfung      19  01003               Lübeck   
2  2020-12-27       18-59  Erstimpfung      13  01053  Herzogtum Lauenburg   
3  2020-12-27         60+  Erstimpfung     200  01053  Herzogtum Lauenburg   
4  2020-12-27       18-59  Erstimpfung      20  01055          Ostholstein   

           Bundesland  
0  Schleswig-Holstein  
1  Schleswig-Holstein  
2  Schleswig-Holstein  
3  Schleswig-Holstein  
4  Schleswig-Holstein

### Prophet 

Für den Forecast muss das DIVI Dataset nochmal angepasst werden. Prophet benötigt für den Forecast zwei Spalten: das Datum und den y-Wert, der vorhergesagt werden soll. Diese beiden Spalten müssen im Datensatz als ds und y hinterlegt sein. 

In [100]:
cases_prophet= divi[['date','faelle_covid_aktuell']].groupby('date').sum().reset_index()
cases_prophet.columns = ['ds','y']

In [101]:
cases_prophet.head()

ds     y
0  2020-04-24  2650
1  2020-04-25  2543
2  2020-04-26  2529
3  2020-04-27  2471
4  2020-04-28  2394

## Predictive Modeling

In [105]:
m = Prophet(interval_width=0.95,changepoint_prior_scale=0.01,seasonality_mode='multiplicative',daily_seasonality=True,weekly_seasonality=True,yearly_seasonality=True,changepoint_range=0.9)
m.fit(cases_prophet)
# Die nächsten 90 Tage werden vorhergesagt
future = m.make_future_dataframe(periods=90)
forecast = m.predict(future)
# Der ursprüngliche y-Wert wird dem Datensatz hinzugefügt.
forecast['y'] = cases_prophet['y']

# Der Forecast wird in einem separaten Datensatz abgespeichert. Es werden nur Zeilen betrachtet, die in der Zukunft liegen und daher keinen y-Wert haben.
pred = forecast[forecast.y.isna()]
pred[['ds','y','yhat']].describe()

Initial log joint probability = -33.3148
    Iter      log prob        ||dx||      ||grad||       alpha      alpha0  # evals  Notes 
      99       1556.16     0.0099208       1200.94      0.8675      0.8675      123   
    Iter      log prob        ||dx||      ||grad||       alpha      alpha0  # evals  Notes 
     167       1620.45   0.000351183       1027.31    5.41e-07       0.001      251  LS failed, Hessian reset 
     199       1625.17    0.00159692        404.82      0.3575           1      292   
    Iter      log prob        ||dx||      ||grad||       alpha      alpha0  # evals  Notes 
     299       1651.03     0.0010482        398.06       0.853       0.853      416   
    Iter      log prob        ||dx||      ||grad||       alpha      alpha0  # evals  Notes 
     399       1668.87    0.00406088       1234.43           1           1      552   
    Iter      log prob        ||dx||      ||grad||       alpha      alpha0  # evals  Notes 
     453       1673.12   0.000269026    

y         yhat
count  0.0    90.000000
mean   NaN  1259.816367
std    NaN   895.866484
min    NaN   230.600608
25%    NaN   534.982713
50%    NaN   898.791357
75%    NaN  1901.822932
max    NaN  3294.127617

Sollte der Forecast negative Fallzahlen vorhersagen, dann werden diese mit einer 0 versehen.
Die Spalte 'yhat' sind die vorhergesagten Fallzahlen. yhat_lower und yhat_upper stellen das Konfidenzintervall dar. 

In [106]:
pred.loc[pred['yhat'] < 0, 'yhat'] = 0
pred.loc[pred['yhat_lower'] < 0, 'yhat_lower'] = 0
pred.loc[pred['yhat_upper'] < 0, 'yhat_upper'] = 0